### PR TITLE
Reduce LDS bank conflicts

### PR DESCRIPTION
**Pull request by giuseros**
Given the `Kouter x D x Kpack` layout of LDS in `rocMLIR`, this PR is implementing an LDS-write bank conflicts reduction policy in  two cases: 
- When the global matrix is `DxK`, this means we would write `Kpack` vectors in the same column (because we are basically transposing the global matrix in LDS). In this case we rotate the colum dimension by the row dimension : i.e., when we want to write at position `(i,j)` in LDS, we get rerouted to position `(i, (i+j)%D)`. 
- When the global matrix is `KxD`, this means that we are writing along the `D` dimension, and that would be fine if every work-item is only writing one single vector. If a workitem is writing >2 vectors, then they are writing in a strided fashion which generates bank conflicts. The solution proposed is to do a coordinate-swap when storing in LDS, so that the contiguous dimension is the workitem id. The output will be wapped back before producing the final result (thanks @manupak for the suggestion!)

These two policies required a bit of code reshuffling:
- I have to propagate the `isKContiguousDim{A/B}` down to the `blockwiseGemm` operator
- I need `copyMPerThread` and `copyMPerThread` in `blockwiseGemm` to address cases where `kpack==1`, this means that `computeCopyPerThread` has been promoted into `loweringUtils.h` because it'll be used by two operations. 
- `wrapLDSBufferForStore` is building the transformation in a top down fashion (instead than bottom up) because I need a transformation like `(d0, d1)->(d0, d1+d0)` which cannot be build bottom up. 

There are at least three big refactors that stem from this PR:
- Moving the `BlockwiseGemmAccel` operation to use transforms (as opposed to plain `AffineLoop`s)
- Wrap the LDS as a view and pass it down, so that we don't need to propagate (or recompute) LDS layout information in `BlockwiseGemmAccel`/`BlockwiseGemm`
- Unify accel and non-accel code paths

### CODE REVIEW

**Comment by giuseros** [comment]
Fixes https://github.com/ROCmSoftwarePlatform/rocMLIR-internal/issues/806

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 1]
```diff
@@ -1155,9 +1144,11 @@ struct GridwiseGemmAccelRewritePattern
 
       // Emit blockwise GEMM.
       blockwiseGemmAccelOp = b.create<BlockwiseGemmAccelOp>(
-          loc, ldsViewForGemmA, ldsViewForGemmB, mMyWaveOffsetA, mMyWaveOffsetB,
-          arrayA, arrayB, regCAllocOp, op.getArchAttr(), op.getFeaturesAttr(),
-          op.getBlockSizeAttr(), op.getParamsAttr());
+          loc, ldsViewForGemmA,
+          BoolAttr::get(op->getContext(), isKContiguousDimA), ldsViewForGemmB,
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: I think there's a builder method for this?

**Review by krzysz00 — COMMENTED** [review]
Very minor comments currently.

I don't see any real issues with the approach, if it works

File: mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp [hunk 2]
```diff
@@ -193,7 +194,9 @@ makeViewsForRowsAndCols(TopDownTMBuilder &viewBuilder, int64_t mPerRepeat,
 
 RegsAsMatrixSubTiles MfmaEmitter::computeOutputTransforms(
     PatternRewriter &b, Location loc, int64_t mLen, int64_t nLen,
-    int64_t blockSize, ArrayRef<int64_t> bidGridLengths) {
+    bool isKContiguousDimA, bool isKContiguousDimB, int64_t computeMPerThread,
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp]: nit-naming : While I agree we do the swap based on K being contigous, might for these utils I;d prefer a more direct command such as 'doSwapThreadIterSubDimsForM' / 'doSwapThreadIterSubDimsForN'.

Also lets not use 'A' and 'B' to make these utils gemm agonstic but more related to an individual matrix, because A & B is only relevant here because we take M from A and N from B. I feel using N & M is more consistent with rest of the code here w/o introducing the concepts of A & B.

File: mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp [hunk 3]
```diff
@@ -305,9 +308,17 @@ RegsAsMatrixSubTiles MfmaEmitter::computeOutputTransforms(
...
     toMatrixC.unmerge("gemmN", 2, dimNamesN, dimSizesN);
-    TransformMapAttr toMatrixCAttr = toMatrixC.get();
-    ret.gridSubTile = b.getArrayAttr(
-        {splitMemoryCoordsAttr, toRowsAndColsAttr, toMatrixCAttr});
+
+    // Before returning the output view, if necessary, swap back the
+    // threadid/iter dimensions on both the M/N axis.
+    SmallVector<Attribute> transformAttrs{splitMemoryCoordsAttr,
+                                          toRowsAndColsAttr};
+    mlir::rock::swapThreadIdAndIteration(
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp:316]: This swap needs to happen for blockwise subTile view as well.

File: mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp [hunk 4]
```diff
@@ -514,7 +525,9 @@ void WmmaEmitter::emitThreadwiseLoop(OpBuilder &b, Location loc, Value argA,
 
 RegsAsMatrixSubTiles WmmaEmitter::computeOutputTransforms(
     PatternRewriter &b, Location loc, int64_t mLen, int64_t nLen,
-    int64_t blockSize, ArrayRef<int64_t> bidGridLengths) {
+    bool isKContiguousDimA, bool isKContiguousDimB, int64_t computeMPerThread,
+    int64_t computeNPerThread, int64_t blockSize,
+    ArrayRef<int64_t> bidGridLengths) {
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp:539]: nothing meaninful have been done with these changes ?
dont we need to do the swap here as well ?

File: mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h [hunk 5]
```diff
@@ -108,6 +109,22 @@ backwardDataKernelIds(int64_t strideHeight, int64_t strideWidth,
...
+/// workitem) reads along K and M (independently on how we vectorize the reads)
+FailureOr<std::pair<int64_t, int64_t>>
+computeCopyPerThread(Type elementType, int64_t copyPerThread, int64_t kPerBlock,
+                     int64_t dPerBlock, int64_t kpack, Location loc);
+
+// if K is not the contiguous dimension, we swapped (on each axis) the thread id
+// and the iter id dimensions, so that the threads write in a contiguous fashion
+// minimizing LDS bank conflicts.  This transformation swap those dimensions
+// back before producing the final output view
+TopDownTMBuilder swapThreadIdAndIteration(
```
Comment (manupak) [inline mlir/include/mlir/Dialect/Rock/utility/loweringUtils.h]: See the comment below on MfmaEmitter::computeOutputTransforms.
I'd like this utility tied to swapping sub dims of 'M' (i.e. rows) and 'N' (i.e. cols) where you should only need
mLen, mPerBlock, mPerThread (same for N). This could only use that information with a more direct naming for the bools doSwapThreadIterSubDimsofM (or similiar).

Lets not introduce concepts of A. B and K here.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 6]
```diff
@@ -185,7 +150,9 @@ computeCopyPerThread(Type elementType, int64_t copyPerThread, int64_t kPerBlock,
 static FailureOr<Value> wrapLDSBufferForStore(OpBuilder &b, Location loc,
                                               Value buffer, Type ldsReadType,
                                               int64_t kOuter, StringRef dName,
-                                              int64_t d, int64_t kPerThread) {
+                                              int64_t d, int64_t kPerThread,
+                                              int64_t dPerThread,
+                                              bool isKContiguousDim) {
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: What if K & D is neither the contiguous dimension ? 
Now I think about it again, I feel we should only do this if global kVecLen > 1.

Nit-naming - id would be better to give a direct name such as 'rotateKwithD' or similar instead of the proxy 'isKContiguousDim' and let the caller decide when to do the rotation.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 7]
```diff
@@ -185,7 +150,9 @@ computeCopyPerThread(Type elementType, int64_t copyPerThread, int64_t kPerBlock,
 static FailureOr<Value> wrapLDSBufferForStore(OpBuilder &b, Location loc,
                                               Value buffer, Type ldsReadType,
                                               int64_t kOuter, StringRef dName,
-                                              int64_t d, int64_t kPerThread) {
+                                              int64_t d, int64_t kPerThread,
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:191]: Also please update the function description as it is misleading now to be said LDS layout is <kOuter * d * kpack * sizeof(T) x i8

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 8]
```diff
@@ -254,6 +276,13 @@ struct BlockwiseGemmRewritePattern
     toLdsIndexA.ignore("n_cuwave");
     toLdsIndexA.passThrough({"kpack"}, {2}, {"kpack"});
     TransformMapAttr toLdsIndexAAttr = toLdsIndexA.get();
+    SmallVector<Attribute, 4> transformAttrsA{splitTidAAttr, toLdsIndexAAttr};
+
+    // Since the LDS layout is `kOuter x rotate(m) x kpack` we want to rotate
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: This needs to be said in the function docs for wrapLDSforStore as well.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 9]
```diff
@@ -242,6 +242,28 @@ struct BlockwiseGemmRewritePattern
...
+
+    Type elementTypeA = blockAType.getElementType();
+    if (auto vectorDataType = elementTypeA.dyn_cast<VectorType>())
+      elementTypeA = vectorDataType.getElementType();
+
+    auto maybeCopyAPerThread = computeCopyPerThread(
+        elementTypeA, aCopyPerThread, k * kPack, m, kPack, loc);
+    if (failed(maybeCopyAPerThread))
+      return maybeCopyAPerThread;
+    int64_t copyMPerThread = (*maybeCopyAPerThread).second;
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: This feels fragile to me...
We shouldn;t be recomputing this as it was decided before the op is created and this coule be problematic if the op is created with a seperate logic.

I;d say we just put an attribute on the op to indicate mPerThread and nPerThread; I think that is all you want to know rather than recomputing it here. It makes the debugging easier in future also less problematic for the second gemm in the attention kernel.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 10]
```diff
@@ -254,6 +276,13 @@ struct BlockwiseGemmRewritePattern
     toLdsIndexA.ignore("n_cuwave");
     toLdsIndexA.passThrough({"kpack"}, {2}, {"kpack"});
     TransformMapAttr toLdsIndexAAttr = toLdsIndexA.get();
+    SmallVector<Attribute, 4> transformAttrsA{splitTidAAttr, toLdsIndexAAttr};
+
+    // Since the LDS layout is `kOuter x rotate(m) x kpack` we want to rotate
+    // the dimension `m` before reading from LDS
+    int64_t strideA = (kPack == 1 ? copyMPerThread : 1);
+    rotateIf(op.getIsKContiguousDimA(), toLdsIndexA, toLdsIndexAAttr, strideA,
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: Please put a cross-reference link to wrapLDSforStore to say that these two needs to be kept in sync.

TLDR;

I spent some time thinking about can we pass the transformMaps of the LDS so these need to not be maintained here.. it feels complex so lets keep a cross-reference here.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 11]
```diff
@@ -423,6 +457,21 @@ struct BlockwiseGemmAccelRewritePattern
     Value mWavesConstantOp = b.create<ConstantIndexOp>(loc, mWaves);
     Value nWavesConstantOp = b.create<ConstantIndexOp>(loc, nWaves);
 
+    int64_t aCopyPerThread = (K * M * KPack) / op.getBlockSize();
+    int64_t bCopyPerThread = (K * N * KPack) / op.getBlockSize();
+
+    auto maybeCopyAPerThread = computeCopyPerThread(dataTypeA, aCopyPerThread,
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: Here also, lets not recompute mPerThread and nPerThread here and pass in as attributes on the basis it was decided when the op is created and it should be upto the caller to decide that.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 12]
```diff
@@ -455,11 +504,31 @@ struct BlockwiseGemmAccelRewritePattern
...
+                                 bool isKContiguousDim,
+                                 int64_t copyMNPerThread) {
       // Compute source offset
       Value sourceOffset = accelEmitterPtr->computeLdsSourceOffset(
           kb, k_i, mnb, mn_i, b, MN, loc, sourceBase, mnWaves, laneId);
 
+      // Since the LDS layou is `k0 x rotate(m) x kpack` we want to rotate
+      // the dimension `n` before reading from LDS. Since we don't use
+      // any transform here, this is done by hand with rem/div operators
+      if (isKContiguousDim) {
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: as before, I think we should be doing this only if kVecLen > 1.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp (hunk 6 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: 'isKContiguousDim' is going to be rather confusing when I call wrapLDSBufferForStore for the seccond gemm. I think 'rotateKwithD' is better because it gives the sense it is optional/perf improvement flag that called can use as long as blockwiseGemm knows that.

 With that I feel its better to name for blockwiseGemm 's new attrbitues to indicate K + D rotation is done on the input rather than saying it was done because K was contigous dim -- which is true but that should only be present in GridwiseGemmtoBlockwise where we make the decision to rotateKwithD when the input tiles for A & B is loaded from global memory.

**Review by manupak — DISMISSED** [review]
Thanks for the PR! Its great to see you have sorted things when K global vectorized scenario.
I m an agreement to binary generated from this PR (I think) and my requested changes are more structural in nature.

I did a detailed review on the PR, -- keeping an eye out specially where coupling is being introduced between gridwise and blockwise lowering is happenning. I wanted to point out places where assumptions being taken how gridwise lowering is happenning when we are doing blockwise lowering. So my main suggestions are to hoist them up to gridwise lowering so that we still allow blockwise gemm to be used in other kernels (Im obviously thinking of the attention kernel).

Also I note wmma computeOutputTransforms are not affected by non-K sub-dim swaps. Is that correct ? Or Is it because we dont have Navi3x in our CI that this is passing ?

Other than that my other comments are naming nits.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp (hunk 9 above)
Comment (giuseros) [reply mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: It is super fragile, but I feared that I was adding too many attributes on the op. But if it is fine with you, I will go ahead and do it :)

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp (hunk 10 above)
Comment (giuseros) [reply mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: The problem is also that on the accel path we don't have transform maps. I would pass the maps once we bring the accel path to use them.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp (hunk 6 above)
Comment (giuseros) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: > Now I think about it again, I feel we should only do this if global kVecLen > 1.

Can we talk more about that? If I understand what is the issue you see, I am not sure kVecLen>1 is either a good condition.

File: mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp (hunk 4 above)
Comment (giuseros) [reply mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp:539]: Yes, I fear that this is because we don't have wmma in our CI. I will try to run the tests on my local card

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp (hunk 12 above)
Comment (giuseros) [reply mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp]: Yeah I "think" I see why you want that, but I am not sure this is the right solution. Resolving this discussion, but leaving the other one open.

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td [hunk 13]
```diff
@@ -1046,8 +1046,12 @@ defvar LdsBufferTypes = GemmInputTypes # [VectorOfRankAndType<[1], GemmInputType
 def Rock_BlockwiseGemmOp:
     Rock_Op<"blockwise_gemm">,
     Arguments<(ins MemRefRankOf<GemmInputTypes, [3]>:$matrixA,
+                   BoolAttr : $isKContiguousDimA,
                    MemRefRankOf<GemmInputTypes, [3]>:$matrixB,
+                   BoolAttr : $isKContiguousDimB,
```
Comment (manupak) [inline mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: as discussed, it might be better to call these rotateKwithD (to denote what exactly they are doing rather than the 'why') and also I think they could be UnitAttr (i.e. optional)

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td [hunk 14]
```diff
@@ -1046,8 +1046,12 @@ defvar LdsBufferTypes = GemmInputTypes # [VectorOfRankAndType<[1], GemmInputType
...
                    MemRefRankOf<GemmOutputTypes, [2]>:$matrixC,
+                   I32Attr:$copyMPerThread,
```
Comment (manupak) [inline mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: nit : we could drop the copy* prefix as it doesn;t need to be copied -- it could also be computed.

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td [hunk 15]
```diff
@@ -1084,7 +1088,11 @@ defvar AccelResTypes = [VectorOfLengthAndType<[4, 8, 16, 32], [F32, I32, F16, BF
 def Rock_BlockwiseGemmAccelOp:
     Rock_Op<"blockwise_gemm_accel">,
     Arguments<(ins MemRefOf<LdsBufferTypes>:$matrixA,
+                   BoolAttr:$isKContiguousDimA,
                    MemRefOf<LdsBufferTypes>:$matrixB,
+                   BoolAttr:$isKContiguousDimB,
+                   I32Attr:$copyMPerThread,
+                   I32Attr:$copyNPerThread,
```
Comment (manupak) [inline mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: Same comment as above

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp (hunk 6 above)
Comment (giuseros) [reply mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: We clarified this point with a chat :) Marking this as resolved

File: mlir/lib/Dialect/Rock/Transforms/AccelEmitter.h [hunk 16]
```diff
@@ -114,7 +114,9 @@ struct AccelEmitter {
   /// matrix multiplication tile.
   virtual RegsAsMatrixSubTiles
   computeOutputTransforms(PatternRewriter &b, Location loc, int64_t mLen,
-                          int64_t nLen, int64_t blockSize,
+                          int64_t nLen, bool isKContiguousDimA,
+                          bool isKContiguousDimB, int64_t computeMPerThread,
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/Transforms/AccelEmitter.h]: naming : I think you missed the changes here.

**Review by manupak — COMMENTED** [review]
I dont have major concerns here assuming Navi3x are working :).

As we discussed offline,
I'd like to gridwise common utils and blockwise gemm to have explicit flags for the rotation (where applicable) and subdim swaps (where applicable) . I can still see KContigousDimA in some places.

We can use the heuristic contiguous dim being actual driver to fix those flags when we do the gridwise to blockwise lowering.

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td (hunk 14 above)
Comment (giuseros) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: The idea here is that the `dPerThread` is how many elements each thread copies from global to LDS, which can be different from how many elements each thread computes. Happy to proceed if you want to still change the name, I just wanted to give you the context of why I picked it

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td (hunk 13 above)
Comment (giuseros) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: In the operator? The point is that we will use that attribute to rotate and/or swap. If we name the operator property `rotateKwithD`, during the lowering we will have:
```
bool rotateDWithK = op.getRotateKWithD
bool doSwapThreadIterSubDimsForM = !rotateDWithK
```
If you wish, we might add a `doSwapThreadIterSubDimsForM` additional property to the blockwise gemm operator to keep things separated. What do you think?

File: mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp (hunk 4 above)
Comment (giuseros) [reply mlir/lib/Dialect/Rock/Transforms/AccelEmitter.cpp:539]: I ran the tests on my navi31 board and they all passed. Marking this as resolved.

**Comment by giuseros** [comment]
@krzysz00 , @manupak I re-ran the tuner and got the performance results, which look quite good (better than I expected). I am attaching the results here:

[ldsPerfComparison.zip](https://github.com/ROCmSoftwarePlatform/rocMLIR/files/12439463/ldsPerfComparison.zip)

Please note that in the original tuning db, entries for 1k,2k4k,8k,  are missing (hence why the `FAILED` rows).

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td (hunk 14 above)
Comment (manupak) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: What I meant to say was the input to blockwiseGemm doesn;t need to come from global -- always.
I think you are talking assumption that input tensors live in global --> regs --> LDS.

```mermaid
flowchart TD
   In1 --> gemm0
   In2 --> gemm0
   gemm0 --> gemm1
   In3 --> gemm1
   gemm1 --> Out1
```

So for `gemm1`'s LHS input data in the thread (dPerThead) needs to be written to LDS. These doesn't match with definition of 'each thread copies from global to LDS'

Comment (manupak) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: That dPerThread is computed not copied.

**Comment by manupak** [comment]
Perf is amazing!

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td (hunk 13 above)
Comment (manupak) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: as discussed offline, I think they are not mutually exclusive from blockwiseGemm PoV.
i.e. it will still result in a valid kernel that is computing accurate output.

They become mutually exclusive in gridwise to blockwise lowering; so lets hoist the decision to the lowering.

Comment (giuseros) [reply mlir/include/mlir/Dialect/Rock/IR/RockOps.td]: That was my bad, because we don't use the swap flag in blockwise, but we only use the rotation flag. So I added the two rotateMWithK and rotateNWithK  as you suggested

**Comment by giuseros** [comment]
I did an analysis of mlir vs rocblas perf, and we actually improved in many cases, so I can confirm there is a good perf improvement, although in same cases it might not be as high as the report I shared. I might still want to do another round of tuning, but perf-wise I think this feature is ready to go

**Comment by giuseros** [comment]
Hi @manupak , is there anything else that needs to be addressed ?

File: mlir/lib/Dialect/Rock/utility/loweringUtils.cpp [hunk 17]
```diff
@@ -344,3 +352,81 @@ FailureOr<RegsAsMatrixSubTiles> mlir::rock::getPackedRegsAsTileViews(
   }
   return gpuViews;
 }
+
+TopDownTMBuilder mlir::rock::swapThreadIdAndIteration(
+    TopDownTMBuilder &toMatrixC, ArrayRef<int64_t> bidGridLengths,
```
Comment (manupak) [inline mlir/lib/Dialect/Rock/utility/loweringUtils.cpp]: bidGridLengths should not be needed here. I mean how the grid is partitioned should not be relevant.
I think what you need is mblocks and nblocks.

File: mlir/include/mlir/Dialect/Rock/IR/RockOps.td [hunk 18]
```diff
@@ -1048,6 +1048,10 @@ def Rock_BlockwiseGemmOp:
     Arguments<(ins MemRefRankOf<GemmInputTypes, [3]>:$matrixA,
                    MemRefRankOf<GemmInputTypes, [3]>:$matrixB,
                    MemRefRankOf<GemmOutputTypes, [2]>:$matrixC,
+                   I32Attr:$inMPerThread,
+                   I32Attr:$inNPerThread,
+                   UnitAttr:$rotateMWithK,
+                   UnitAttr:$rotateNWithK,
```
Comment (manupak) [inline mlir/include/mlir/Dialect/Rock/IR/RockOps.td:1054]: We might need to describe these flags in the description (below) for others who haven;t followed this PR.

**Review by manupak — COMMENTED** [review]
almost there... (Im droping my hold as comments are minor)

File: mlir/lib/Dialect/Rock/utility/loweringUtils.cpp (hunk 17 above)
Comment (manupak) [reply mlir/lib/Dialect/Rock/utility/loweringUtils.cpp]: to clarify further
bidGridLengths[1] == mblocks and  bidGridLengths[2] == nblocks will not always be the case..

**Comment by manupak** [comment]
> I did an analysis of mlir vs rocblas perf, and we actually improved in many cases, so I can confirm there is a good perf improvement, although in same cases it might not be as high as the report I shared. I might still want to do another round of tuning, but perf-wise I think this feature is ready to go

Since this is a purely a perf related PR, we might want to attach some numbers here.
IIRC, the early data was not relevant because the tuning was not *fair* ? 

If getting tuning data is the problematic aspect,
I feel we could do a iso-tuning (using same tuning db) perf comparison -- taking the tuning db from develop (i.e. weeklies)

**Comment by manupak** [comment]
Code-wise approved! Thanks for bearing with me...
(just waiting on to see the perf diff)

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 19]
```diff
@@ -143,9 +143,9 @@ bestGlobalVectorization(OpBuilder &b, Value matrix, int64_t copyDPerThread,
 
 /// Compute a thread copy layout, i.e., how many elements a single thread (or
 /// workitem) reads along K and M (independently on how we vectorize the reads)
-static FailureOr<std::pair<int64_t, int64_t>>
-computeCopyPerThread(Type elementType, int64_t copyPerThread, int64_t kPerBlock,
-                     int64_t dPerBlock, int64_t kpack, Location loc) {
+FailureOr<std::pair<int64_t, int64_t>> static computeCopyPerThread(
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: Why'd you move the `static`?

File: mlir/include/mlir/Dialect/Rock/utility/transformMapUtils.h [hunk 20]
```diff
@@ -136,6 +137,16 @@ TransformMapAttr transformExtractSlice(OpBuilder &b, Location loc,
...
 
+// If the condition is satified, rotate the dimension `d` by `k` using
+// `d = (d+k*stride) % len(d)`
+rock::TopDownTMBuilder rotateIf(bool condition, TopDownTMBuilder &builder,
+                                TransformMapAttr &attr, int64_t stride,
+                                StringRef dName, int64_t d, int64_t dPos,
+                                StringRef kName, int64_t k,
+                                ArrayRef<StringRef> beforeDims,
+                                ArrayRef<StringRef> afterDims,
+                                SmallVector<Attribute, 4> &transformAttrs);
```
Comment (krzysz00) [inline mlir/include/mlir/Dialect/Rock/utility/transformMapUtils.h]: Nit: `SmallVectorImpl<Attribute>&` so you don't need to specify the size.

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 21]
```diff
@@ -176,16 +176,19 @@ computeCopyPerThread(Type elementType, int64_t copyPerThread, int64_t kPerBlock,
...
 }
 
-/// Wraps the LDS buffer "buffer", which is <kOuter * d * kpack * sizeof(T) x i8
-/// into a tid x iter view, where `iter` iterates over nominal scalar indices
-/// into a buffer of type T. `buffer` will be reinterpreted as a buffer with
-/// element type vector<kpackPerThread x T> (with kpackPerThread == 1 meaning
-/// just T). The resulting view must be iterated over with a stride of no less
-/// than min(kPerThread, kpack).
+/// Wraps the LDS buffer "buffer", which is <kOuter * rotate(d) * kpack *
+/// sizeof(T) x i8 into a tid x iter view, where `iter` iterates over nominal
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp]: Comment nit: I woludn't put the rotate in here, and add back the bracket after i8, to emphasize that the input is a 1D buffer.

Then I'd mention that the `d` dimension is rotated in a separate sentence

**Review by krzysz00 — APPROVED** [review]
Looks reasonable, I've mainly got nitpicks.

Do you know where some of those -7% drops in the perf report come from? Are they still there on a rerun (using the same perf configs)?

**Comment by giuseros** [comment]
Thanks @krzysz00 ! About the perf degrations, I am not sure. I can reproduce them, and it seems only to happen in `int8` world. I can see the LDS bank conflicts going down, but maybe we were relying on the loads to take time to balance between different threadd

**Comment by krzysz00** [comment]
Re the int8 degradations, if they don't show up consistently, then it's probably fine.

**Comment by giuseros** [comment]
Ok, some news about this. Even with rocm-head data, the perf seemed skewed, so I bit the bullet and ran another tuning session over the week-end. The result was that we were having performance regressions in int8, ranging between -8%  up to -15%. 

I investigated a bit, and the problem is that the tricks that we are using to improve bank conflicts might cause troubles during vectorization. 

E.g., if if maxVecLen=16 and kpack=4, each thread is writing 4 contiguous vectors (each of 4 elements) in LDS. LLVM can vectorize this by a single ds_write_128. However, if we swap thread-id and iter-id dimensions we will loose this vectorization opportunity.

So I added some more conditions to decide when to rotate/swap dims. This removed the int8 regressions. 

Attaching here the new report (which can considered final)

[final.zip](https://github.com/ROCmSoftwarePlatform/rocMLIR/files/12512957/final.zip)
